### PR TITLE
Remove `StructuredDiag`

### PR DESCRIPTION
**Pull request by oli-obk**
follow-up to https://github.com/rust-lang/rust/pull/127319

This trait was an experiment that didn't pan out.

### CODE REVIEW

File: compiler/rustc_hir_analysis/src/errors/wrong_number_of_generic_args.rs [hunk 1]
```diff
@@ -1120,17 +1117,16 @@ impl<'a, 'tcx> WrongNumberOfGenericArgs<'a, 'tcx> {
...
-
-    fn diagnostic_common(&self) -> Diag<'tcx> {
-        let mut err = self.start_diagnostics();
+impl<'a, G: EmissionGuarantee> Diagnostic<'a, G> for WrongNumberOfGenericArgs<'_, '_> {
+    fn into_diag(
+        self,
+        dcx: rustc_errors::DiagCtxtHandle<'a>,
+        level: rustc_errors::Level,
+    ) -> Diag<'a, G> {
+        let msg = self.create_error_message();
```
Comment (oli-obk) [inline compiler/rustc_hir_analysis/src/errors/wrong_number_of_generic_args.rs:1126]: This needs to use translatable diagnostics, but that's something to do in a separate PR

File: compiler/rustc_hir_typeck/messages.ftl [hunk 2]
```diff
@@ -138,6 +139,11 @@ hir_typeck_option_result_asref = use `{$def_path}::as_ref` to convert `{$expecte
 hir_typeck_option_result_cloned = use `{$def_path}::cloned` to clone the value inside the `{$def_path}`
 hir_typeck_option_result_copied = use `{$def_path}::copied` to copy the value inside the `{$def_path}`
 
+hir_typeck_pass_to_variadic_function = can't pass `{$ty}` to variadic function
```
Comment (workingjubilee) [inline compiler/rustc_hir_typeck/messages.ftl:142]: C calls them the _variable arguments_.
```suggestion
hir_typeck_pass_to_variadic_function = can't pass `{$ty}` into variable arguments
```

File: compiler/rustc_hir_typeck/messages.ftl [hunk 3]
```diff
@@ -138,6 +139,11 @@ hir_typeck_option_result_asref = use `{$def_path}::as_ref` to convert `{$expecte
...
+    .suggestion = cast the value to `{$cast_ty}`
+    .help = cast the value to `{$cast_ty}`
+    .teach_help = certain types, like `{$ty}`, must be casted before passing them to a variadic function, because of arcane ABI rules dictated by the C standard
```
Comment (workingjubilee) [inline compiler/rustc_hir_typeck/messages.ftl:145]: It is not an ABI detail, it is a source language requirement: passing the argument triggers a _syntactic_ argument promotion.
```suggestion
    .teach_help = a `{$ty}` would be cast to `{$cast_ty}` by C's default argument promotion
```

**Review by workingjubilee — COMMENTED** [review]
not required I guess

File: tests/ui/error-codes/E0617.rs [hunk 4]
```diff
@@ -5,22 +5,22 @@ extern "C" {
 fn main() {
     unsafe {
         printf(::std::ptr::null(), 0f32);
-        //~^ ERROR can't pass `f32` to variadic function
+        //~^ ERROR can't pass `f32` into variable arguments
```
Comment (compiler-errors) [inline tests/ui/error-codes/E0617.rs:8]: @workingjubilee yes C calls it "variable arguments" but I think it still calls the functions "variadic functions". I think the latter is a more unique phrase and we should keep it. Do you have an example of an error message that calls them "variable arguments"? I searched and couldn't really find anything mentioning either, since C error messages are kinda terse, but I did find this one that calls them "variadic functions":

https://godbolt.org/z/9c79z9j9Y

```
<source>:10:12: error: cannot pass object of non-trivial type 'std::string' (aka 'basic_string<char>') through variadic function; call will abort at runtime [-Wnon-pod-varargs]
    foo(1, s, 3);
           ^
1 error generated.
Compiler returned: 1
```

File: compiler/rustc_hir_analysis/src/structured_errors/wrong_number_of_generic_args.rs
[the remark this answers is not in the document]
Comment (fmease) [reply compiler/rustc_hir_analysis/src/structured_errors/wrong_number_of_generic_args.rs:1]: Actually, as far as I can tell `WrongNumberOfGenericArgs` is only ever emitted at `Level::Error`. Did I miss a usage site? We could make all those methods and the `Diagnostic` impl non-generic.

**Review by fmease — APPROVED** [review]
r=me after addressing my comment in one way or another as well as err's comment.

File: compiler/rustc_hir_analysis/src/structured_errors/wrong_number_of_generic_args.rs
[the remark this answers is not in the document]
Comment (oli-obk) [reply compiler/rustc_hir_analysis/src/structured_errors/wrong_number_of_generic_args.rs:1]: I initially had that, but then remembered something about trying to avoid doing that and instead having all impls be generic (and in fact there are no non-generic impls)

File: tests/ui/error-codes/E0617.rs (hunk 4 above)
Comment (workingjubilee) [reply tests/ui/error-codes/E0617.rs:8]: A niladic function would be one that takes zero arguments, a monadic function, one, a dyadic function, two, etc., by extension we arrive at variadic functions for a variable number of arguments. A function using C's `...` is indeed a variadic function. However, your example function will be called with 1 fixed argument and an unknown number of variable arguments. It is still a variadic function if no variable arguments are passed at all. It can, in those fixed parameters, accept arguments of types prohibited in the variable arguments. Differing from C's terminology seems fine, but the previous message was incorrect in any case.

I realize this is extremely pedantic, but we really ought to be at least a few notches more pedantic about how we handle varargs.

The preference in C/C++ compilers handling erroneous cases seems to be literally just calling it `...` which also seems fine, since that's what the programmer will actually be looking at. https://godbolt.org/z/81nEEEnjo

```
In file included from <source>:3:
<source>: In function 'void foo(int, ...)':
<source>:8:26: error: 'float' is promoted to 'double' when passed through '...' [-Werror]
    8 |     float f = va_arg(ap, float);
      |                          ^
<source>:8:26: note: (so you should pass 'double' not 'float' to 'va_arg')
<source>:8:26: note: if this code is reached, the program will abort
cc1plus: all warnings being treated as errors
Compiler returned: 1
```

Comment (fmease) [reply tests/ui/error-codes/E0617.rs:8]: To be fair, ``"can't pass `$ProblematicType` to variadic function"`` is imprecise (*incorrect* even, technically speaking) since it's legal to pass a value of $ProblematicType as argument where the correp. param is "non-variable". E.g. `0f32` in `f(0f32, x, y, z)` is fine given e.g.`fn f(_: f32, ...)`.

Still, the phrasing does feel a bit uncommon.

Comment (workingjubilee) [reply tests/ui/error-codes/E0617.rs:8]: ...and this is an important distinction to preserve because some C ABIs are written in such a way that the way the arguments will be inspected in the callee function will be handled uniformly regardless of the number of arguments. In other words, they make it so `...` is ABI-compatible with a hypothetical function with the same fixed arguments. So they do in fact deliberately conflate "variadic function" with "a function receiving any number of arguments", because the ABIs are identical.

**But others are not**, and the way varargs are passed has nothing to do with the way other args are passed, so this is in fact completely incompatible. On those platforms, code which conflates these will explode, and some C libraries had to be rewritten for those platforms.

File: compiler/rustc_hir_typeck/messages.ftl [hunk 5]
```diff
@@ -138,6 +139,11 @@ hir_typeck_option_result_asref = use `{$def_path}::as_ref` to convert `{$expecte
 hir_typeck_option_result_cloned = use `{$def_path}::cloned` to clone the value inside the `{$def_path}`
 hir_typeck_option_result_copied = use `{$def_path}::copied` to copy the value inside the `{$def_path}`
 
+hir_typeck_pass_to_variadic_function = can't pass `{$ty}` into variable arguments
```
Comment (workingjubilee) [inline compiler/rustc_hir_typeck/messages.ftl:142]: some options:
```suggestion
hir_typeck_pass_to_variadic_function = can't pass `{$ty}` into variadic arguments
```

```suggestion
hir_typeck_pass_to_variadic_function = can't pass `{$ty}` into `...`
```

Comment (compiler-errors) [reply compiler/rustc_hir_typeck/messages.ftl:142]: prefer either of these 👍

Comment (fmease) [reply compiler/rustc_hir_typeck/messages.ftl:142]: In any case (esp. if we were to take the "`` `...` ``" one, we might also want to add a secondary highlight to the def site of the CVarArgs with a label to clear up any potential confusion.

**Comment by fmease** [comment]
Oli, feel free to throw the user-visible changes to E0617 out of this PR and into a new one. I don't want to block this refactoring PR on diagnostic changes.
